pandas/core/groupby.py

```diff
@@ -3740,9 +3740,31 @@ def _reindex_output(self, result):
             return result
 
         levels_list = [ping.group_index for ping in groupings]
-        index = MultiIndex.from_product(levels_list, names=self.grouper.names)
-        d = {self.obj._get_axis_name(self.axis): index, 'copy': False}
-        return result.reindex(**d).sortlevel(axis=self.axis)
+        index, _ = MultiIndex.from_product(
+            levels_list, names=self.grouper.names).sortlevel()
+
+        if self.as_index:
+            d = {self.obj._get_axis_name(self.axis): index, 'copy': False}
+            return result.reindex(**d)
+
+        # GH 13204
```

Comment:
OK, it can probably be done but there's one issue here: if we have a non-column grouper, should it become a column of the final `result`? This will happen after `reset_index()`. And its name can clash with a name of some other column.